### PR TITLE
data: ship "snap.mount" service that ensures /snap is MS_SHARED

### PR DESCRIPTION
**Pull request by mvo5**
This also cleansup the 14.04 packaging which is using packaging-14.04/debian/*.service but now this is all available in data/systemd so we can drop the duplicated content. Lets see if tests agree :)

LP: #1668659

### CODE REVIEW

**Comment by pedronis** [comment]
conflicts

File: packaging/ubuntu-16.04/rules [hunk 1]
```diff
@@ -108,6 +108,10 @@ override_dh_systemd_enable:
 	dh_systemd_enable \
 		-psnapd \
 		data/systemd/snapd.autoimport.service
+	# mount workaround
```
Comment (zyga) [inline packaging/ubuntu-16.04/rules]: Hmm, we really don't want this on 16.04

**Comment by stgraber** [comment]
@zyga The point of this work is to make /snap a bind-mount everywhere, that's especially import for Ubuntu 16.04 containers which do use systemd. So this isn't work that's meant just for the sake of non-systemd systems.

**Comment by stgraber** [comment]
It's an absolutely valid setup for someone to have their / not be marked as rshared, no matter the init system, and snapd fails to run on any of those, be them containers, vms or physical machines.

**Comment by niemeyer** [comment]
The above comments are implying a lot without much detailing of why they do so, on both ends. Can we please be a bit more explicit and descriptive on the rationale?

@stgraber Why do we need the entries to be shared and not rshared (why does LXD need it that way?).

@mvo5 Can we please include @stgraber's answer to the above question as a comment in the code?

@zyga Why do you think the logic isn't reasonable for 16.04?  Why was it rshared in the first place, and why do we want to keep it that way?

**Comment by zyga** [comment]
@niemeyer I assumed that this was just added by accident, since we unified 14.04 packaging where it did play a part. Given the use for containers on 16.04 I agree that this is valid but I think we should check why things fail and if the mount unit is properly accounted for by our 16.04 tests.

**Comment by zyga** [comment]
@niemeyer 16.04 has everything `rshared` by default and we rely on this when additional snaps are mounted (otherwise such mount events don't propagate into the snap namespace)

**Comment by zyga** [comment]
@niemeyer since systemd defaults to having `rshare` / we didn't need to do it manually but in certain environments (containers) and without full-blown systemd (14.04) this is not the case.

**Comment by stgraber** [comment]
@niemeyer shared and rshared are the same thing, it just indicates whether to change any existing mount under the mountpoint too. Since we're overriding the mount-point with a bind-mount, there's no point in using rshared in this case. (rshared makes sense when doing a remount of an existing mountpoint)

For the rest, @zyga is right, the current code assumes that / will always be rshared, which tends to be true on systemd systems, except for containers and except for some cases where / is mounted from the initrd with odd mount flags (nfs root, iscsi, ...) or when the user manually sets their / to not be rshared (that used to be very popular with packagers as schroot/sbuild would break on rshared systems).

It's really best not to make any assumption when it comes to mount propagation and just ensure that your setup will be the same everywhere for the paths you care about.

**Comment by stgraber** [comment]
@zyga @mvo5 any update on this? I just got bitten by this issue recently where a number of prometheus containers stopped working overnight because the snap got refreshed and failed to restart...

**Comment by mvo5** [comment]
@stgraber I looked into this again and it seems like the mount unit is a bit fragile. It fails during the tests when e.g. snapd is removed:
```
+ dpkg --purge snapd
(Reading database ... 73384 files and directories currently installed.)
Removing snapd (1337.2.23.6) ...
Job for snap.mount failed. See "systemctl status snap.mount" and "journalctl -xe" for details.
dpkg: error processing package snapd (--purge):
 subprocess installed pre-removal script returned error exit status 1
Errors were encountered while processing:
 snapd
```
I wonder if instead we should make this part of snapd itself, i.e. on startup do: `mount -o remount,shared /snap` ?

**Comment by mvo5** [comment]
Closing for now, I think we need a different approach here.

**Comment by zyga** [comment]
@mvo5 note that it must be shared before the first snap app runs or that app will misbehave. I agree we want to do it elsewhere, perhaps in snap-confine itself? There's a global initialization we already do (holding a lock) that could be used to initialize this part reliably as well. We already have code for checking mountinfo so we can grab the lock, see if /snap needs sharing and just do it.

**Comment by mvo5** [comment]
Superseded by https://github.com/snapcore/snapd/pull/3136

**Pull request closed without merging**